src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioDiagnosticListTable.cs

```diff
@@ -130,5 +150,98 @@ private void OnErrorListPropertyChanged(object sender, PropertyChangedEventArgs
                 AddTableSourceIfNecessary(this.Workspace.CurrentSolution);
             }
         }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            Contract.ThrowIfFalse(_errorList is IErrorList2);
+
+            switch (e.Kind)
+            {
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.SolutionChanged:
+                case WorkspaceChangeKind.SolutionCleared:
+                case WorkspaceChangeKind.SolutionReloaded:
+                case WorkspaceChangeKind.SolutionRemoved:
+                case WorkspaceChangeKind.ProjectAdded:
+                case WorkspaceChangeKind.ProjectChanged:
+                case WorkspaceChangeKind.ProjectRemoved:
+                    // Set error list toggle state based on current analysis state for all languages for projects in current solution.
+                    var fullAnalysisState = _optionService.GetOption(RuntimeOptions.FullSolutionAnalysis);
+                    if (fullAnalysisState)
+                    {
+                        var langauges = e.NewSolution.Projects.Select(p => p.Language).Distinct();
+                        foreach (var language in langauges)
+                        {
+                            if (!ServiceFeatureOnOffOptions.IsClosedFileDiagnosticsEnabled(_optionService, language))
+                            {
+                                fullAnalysisState = false;
+                                break;
+                            }
+                        }
+                    }
+
+                    ((IErrorList2)_errorList).AnalysisToggleState = fullAnalysisState;
+                    return;
+            }
+        }
+
+        private void OnErrorListFullSolutionAnalysisToggled(object sender, AnalysisToggleStateChangedEventArgs e)
+        {
+            var newOptions = _optionService.GetOptions()
+                .WithChangedOption(RuntimeOptions.FullSolutionAnalysis, e.NewState)
+                .WithChangedOption(ServiceFeatureOnOffOptions.ClosedFileDiagnostic, LanguageNames.CSharp, e.NewState)
+                .WithChangedOption(ServiceFeatureOnOffOptions.ClosedFileDiagnostic, LanguageNames.VisualBasic, e.NewState)
+                .WithChangedOption(ServiceFeatureOnOffOptions.ClosedFileDiagnostic, TypeScriptLanguageName, e.NewState);
+            
+            _optionService.SetOptions(newOptions);
+        }
+
+        private void OnOptionChanged(object sender, OptionChangedEventArgs e)
+        {
+            Contract.ThrowIfFalse(_errorList is IErrorList2);
+
+            if (e.Option == RuntimeOptions.FullSolutionAnalysis || e.Option == ServiceFeatureOnOffOptions.ClosedFileDiagnostic)
```

Comment:
OnWorkspaceChanged is handling the scenario where current set of project languages in solution can change - we are attempting to find if there is at least one project in solution with FSD disabled, in which case error list button will be FSD off. 
This method is handling a single option changed event and if disabled telling the error list to be turned off. We don't do anything for the reverse (option being turned on) as their might be other error list providers (say TS) that still have FSD off.